delivery/common/jwt/examples/jwt-hs256/jwt.js

```diff
@@ -0,0 +1,128 @@
+
+class JWTUtil {
+    static isEmptyString(str) {
+        return !str || 0 === str.trim().length;
+    }
+}
+
+class JWTValidator {
+    constructor(jwtOptions) {
+        this.jwtOptions = jwtOptions || {}, this.algorithms = [ "NONE", "RS256", "RS384", "RS512", "HS256", "HS384", "HS512", "PS256", "PS384", "PS512", "ES256", "ES384", "ES512" ], 
+        this.validateOptionTypes();
+    }
+    async validate(base64JWTToken, keys) {
+        var _a;
+        if ("string" != typeof base64JWTToken) throw new Error("Invalid token type, expected string!");
+        if (!Array.isArray(keys) || !keys.every((elem => void 0 !== elem.type || void 0 !== elem.extractable || null != elem.algorithm || null != elem.usages))) throw new Error("Invalid keys type, expected list of CryptoKey!");
+        const jwtParts = base64JWTToken.split(".");
+        if (jwtParts.length > 3 || jwtParts.length < 2) throw new Error("JWT malformed: Invalid number of parts for JWT token. expected 3 or 2 (unsecured JWT)!");
+        if (JWTUtil.isEmptyString(jwtParts[0])) throw new Error("JWT malformed: jwt header cannot be empty");
+        if (JWTUtil.isEmptyString(jwtParts[1])) throw new Error("JWT malformed: jwt payload cannot be empty");
+        const jwtHBin = atob(jwtParts[0]), jwtPBin = atob(jwtParts[1]), jwtHeader = JSON.parse(jwtHBin), jwtPayload = JSON.parse(jwtPBin);
+        if (this.jwtOptions.issuer && jwtPayload.iss && this.jwtOptions.issuer !== jwtPayload.iss) throw new Error(`JWT malformed: invalid iss, expected ${this.jwtOptions.issuer}`);
+        if (this.jwtOptions.subject && jwtPayload.sub && this.jwtOptions.subject !== jwtPayload.sub) throw new Error(`JWT malformed: invalid sub, expected ${this.jwtOptions.subject}`);
+        if (this.jwtOptions.audience) {
+            const aud = jwtPayload.aud;
+            if (aud && (Array.isArray(aud) || "string" == typeof aud)) {
+                if (!(Array.isArray(aud) ? aud : [ aud ]).includes(this.jwtOptions.audience)) throw new Error(`JWT malformed: invalid aud, expected ${this.jwtOptions.audience}`);
+            }
+        }
+        const clockTimestamp = Math.floor(Date.now() / 1e3);
+        if (!1 === this.jwtOptions.ignoreExpiration) {
+            if ("number" != typeof jwtPayload.exp) throw new Error("JWT malformed: exp must be number");
+            if (jwtPayload.exp && clockTimestamp > jwtPayload.exp + (this.jwtOptions.clockTolerance || 0)) throw new Error("JWT expired");
+        }
+        if (!1 === this.jwtOptions.ignoreNotBefore) {
+            if ("number" != typeof jwtPayload.nbf) throw new Error("JWT malformed: nbf must be number");
+            if (jwtPayload.nbf && jwtPayload.nbf > clockTimestamp + (this.jwtOptions.clockTolerance || 0)) throw new Error("JWT not active");
+        }
+        if (!jwtHeader.alg) throw new Error("JWT malformed: expected alg field in JWT header");
+        if ("NONE" === jwtHeader.alg.toUpperCase()) return {
```

Comment:
Doesn't this mean an attacker could opt out of the JWT validation by supplying an `alg` parameter of `NONE`?

Shouldn't the caller indicate that `NONE` is an acceptable value instead?

Reply:
Sure, will provide an option for the caller to enable NONE as acceptable alg. By default NONE is not supported.

Reply:
Updated with the mentioned changes